src/home/homework-cifs.c

```diff
@@ -53,49 +54,51 @@ int home_setup_cifs(
 
+        if (asprintf(&options, "user=%s,uid=" UID_FMT ",forceuid,gid=" GID_FMT ",forcegid,file_mode=0%3o,dir_mode=0%3o",
+                     user_record_cifs_user_name(h), h->uid, user_record_gid(h), user_record_access_mode(h),
+                     user_record_access_mode(h)) < 0)
+                return log_oom();
+
+        if (h->cifs_domain)
+                if (!strextendf_with_separator(&options, ",", "domain=%s", h->cifs_domain))
+                        return log_oom();
+
+        if (h->cifs_extra_mount_options)
+                if (!strextend_with_separator(&options, ",", h->cifs_extra_mount_options))
+                        return log_oom();
+
         r = home_unshare_and_mkdir();
         if (r < 0)
                 return r;
 
         STRV_FOREACH(pw, h->password) {
-                _cleanup_(unlink_and_freep) char *p = NULL;
-                _cleanup_free_ char *options = NULL;
-                _cleanup_fclose_ FILE *f = NULL;
+                _cleanup_close_ int passwd_fd = -EBADF;
                 pid_t mount_pid;
                 int exit_status;
 
-                r = fopen_temporary_child(NULL, &f, &p);
-                if (r < 0)
-                        return log_error_errno(r, "Failed to create temporary credentials file: %m");
-
-                fprintf(f,
-                        "username=%s\n"
-                        "password=%s\n",
-                        user_record_cifs_user_name(h),
-                        *pw);
-
-                if (h->cifs_domain)
-                        fprintf(f, "domain=%s\n", h->cifs_domain);
-
-                r = fflush_and_check(f);
-                if (r < 0)
-                        return log_error_errno(r, "Failed to write temporary credentials file: %m");
-
-                f = safe_fclose(f);
-
-                if (asprintf(&options, "credentials=%s,uid=" UID_FMT ",forceuid,gid=" GID_FMT ",forcegid,file_mode=0%3o,dir_mode=0%3o",
-                             p, h->uid, user_record_gid(h), user_record_access_mode(h), user_record_access_mode(h)) < 0)
-                        return log_oom();
-
-                if (h->cifs_extra_mount_options)
-                        if (!strextend_with_separator(&options, ",", h->cifs_extra_mount_options))
-                                return log_oom();
+                passwd_fd = acquire_data_fd(*pw, strlen(*pw), /* flags= */ 0);
+                if (passwd_fd < 0)
+                        return log_error_errno(passwd_fd, "Failed to create data FD for password: %m");
 
                 r = safe_fork("(mount)", FORK_RESET_SIGNALS|FORK_RLIMIT_NOFILE_SAFE|FORK_DEATHSIG_SIGTERM|FORK_LOG|FORK_STDOUT_TO_STDERR, &mount_pid);
                 if (r < 0)
                         return r;
                 if (r == 0) {
                         /* Child */
+
+                        char buf[DECIMAL_STR_MAX(passwd_fd)];
+                        sprintf(buf, "%d", passwd_fd);
+                        if (setenv("PASSWD_FD", buf, 1) < 0) {
```

Comment:
one of these days we should probably introduce setenvf() that combines asprintf() and setenv()...

but don't bother with that here...

Reply:
It exists: https://github.com/systemd/systemd/blob/main/src/basic/env-util.c#L1132

I can use it if you'd preferer